server/parsec/components/postgresql/invite.py

```diff
@@ -1097,55 +1374,563 @@ async def test_dump_all_invitations(
+            invitation_created_by_email,
+            invitation_created_by_label,
+            invitation_claimer_email,
+            invitation_created_on,
+            invitation_deleted_on,
+            invitation_deleted_reason,
+            greeter,
+            greeting_attempt_claimer_joined,
+            greeting_attempt_greeter_joined,
+            greeting_attempt_cancelled_by,
+            greeting_attempt_cancelled_reason,
+            greeting_attempt_cancelled_on,
+        ) = row
+        invitation_deleted_reason = (
+            InvitationStatus.from_str(invitation_deleted_reason)
+            if invitation_deleted_reason is not None
+            else None
+        )
+        invitation_info = InvitationInfo(
+            internal_id=invitation_internal_id,
+            type=InvitationType.from_str(invitation_type),
+            created_by_user_id=UserID.from_hex(invitation_created_by_user_id_str),
+            created_by_device_id=DeviceID.from_hex(invitation_created_by_device_id_str),
+            created_by_email=invitation_created_by_email,
+            created_by_label=invitation_created_by_label,
+            claimer_email=invitation_claimer_email,
+            created_on=invitation_created_on,
+            deleted_on=invitation_deleted_on,
+            deleted_reason=invitation_deleted_reason,
+        )
+        greeting_attempt_cancelled_reason = (
+            CancelledGreetingAttemptReason.from_str(greeting_attempt_cancelled_reason)
+            if greeting_attempt_cancelled_reason is not None
+            else None
+        )
+        greeting_attempt_cancelled_by = (
+            GreeterOrClaimer.from_str(greeting_attempt_cancelled_by)
+            if greeting_attempt_cancelled_by is not None
+            else None
+        )
+        return GreetingAttemptInfo(
+            internal_id=greeting_attempt_internal_id,
+            greeting_attempt_id=GreetingAttemptID.from_hex(greeting_attempt_id),
+            invitation_info=invitation_info,
+            greeter=UserID.from_hex(greeter),
+            claimer_joined=greeting_attempt_claimer_joined,
+            greeter_joined=greeting_attempt_greeter_joined,
+            cancelled_by=greeting_attempt_cancelled_by,
+            cancelled_reason=greeting_attempt_cancelled_reason,
+            cancelled_on=greeting_attempt_cancelled_on,
+        )
+
+    async def cancel_greeting_attempt(
+        self,
+        conn: AsyncpgConnection,
+        greeting_attempt_internal_id: int,
+        origin: GreeterOrClaimer,
+        reason: CancelledGreetingAttemptReason,
+        now: DateTime,
+    ) -> None:
+        await conn.execute(
+            *_q_cancel_greeting_attempt(
+                greeting_attempt_internal_id=greeting_attempt_internal_id,
+                cancelled_by=origin.str,
+                cancelled_reason=reason.str,
+                cancelled_on=now,
+            )
+        )
+
+    class StepOutcome(Enum):
```

Comment:
Why is StepOutcome defined inside PGInviteComponent ?

Reply:
To be closer to its usage